spec/cucumber/formatter/json_spec.rb

```diff
@@ -0,0 +1,640 @@
+                "uri": "spec.feature",
+                "keyword": "Feature",
+                "name": "Banana party",
+                "line": 1,
+                "description": "",
+                "elements":
+                 [{"id": "banana-party;monkey-eats-bananas",
+                   "keyword": "Scenario",
+                   "name": "Monkey eats bananas",
+                   "line": 3,
+                   "description": "",
+                   "type": "scenario",
+                   "steps":
+                    [{"keyword": "Given ",
+                      "name": "there are bananas",
+                      "line": 4,
+                      "embeddings": [{"mime_type": "image/png",
+                                      "data": "Zm9v"}],
+                      "match": {"location": "spec/cucumber/formatter/json_spec.rb:500"},
+                      "result": {"status": "passed",
+                                 "duration": 1}}]}]}]})
+          end
+        end
+
+        describe "with a scenario with hooks" do
+          define_feature <<-FEATURE
+          Feature: Banana party
+
+            Scenario: Monkey eats bananas
+              Given there are bananas
+            FEATURE
+
+          define_steps do
+            Before() {}
+            Before() {}
+            After() {}
+            After() {}
+            AfterStep() {}
+            AfterStep() {}
+            Around() { |scenario, block| block.call }
+            Given(/^there are bananas$/) {}
+          end
+
+          it "includes all hooks except the around hook in the json data" do
+            expect(load_normalised_json(@out)).to eq MultiJson.load(%{
+              [{"id": "banana-party",
+                "uri": "spec.feature",
+                "keyword": "Feature",
+                "name": "Banana party",
+                "line": 1,
+                "description": "",
+                "elements":
+                 [{"id": "banana-party;monkey-eats-bananas",
+                   "keyword": "Scenario",
+                   "name": "Monkey eats bananas",
+                   "line": 3,
+                   "description": "",
+                   "type": "scenario",
+                   "before":
+                    [{"match": {"location": "spec/cucumber/formatter/json_spec.rb:545"},
+                      "result": {"status": "passed",
+                                 "duration": 1}},
+                     {"match": {"location": "spec/cucumber/formatter/json_spec.rb:546"},
+                      "result": {"status": "passed",
+                                 "duration": 1}}],
+                   "steps":
+                    [{"keyword": "Given ",
+                      "name": "there are bananas",
+                      "line": 4,
+                      "match": {"location": "spec/cucumber/formatter/json_spec.rb:552"},
+                      "result": {"status": "passed",
+                                 "duration": 1},
+                      "after":
```

Comment:
Specific question 5, the json output is generally made consistent with what Cucumber-JVM produces, but here is uncharted territory, since Cucumber-JVM does not have AfterStep-hooks. I choose to model them as "after" hooks applied to "steps" (so that the three hook types are "before", "after" and "around", and that these types can be applied on different levels - well so far only one hook is applied to something other the a scenario/test case, the `AfterStep()` hook)